AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/helper/ValidatorParameterHelper.java

```diff
@@ -41,89 +39,105 @@
-		T extendsType(String fullyQualifiedName);
-
-		T type(String fullyQualifiedName);
+	public class NoParamValidator implements Validator {
 
-		T anyType();
+		@Override
+		public void validate(ExecutableElement executableElement, IsValid valid) {
+			if (!executableElement.getParameters().isEmpty()) {
+				annotationHelper.printAnnotationError(executableElement, "%s cannot have any parameters");
+				valid.invalidate();
+			}
+		}
 	}
 
-	public static class ParameterRequirement<O extends Validator> implements Validator {
-
-		protected final ParameterValidator<?> validator;
+	public class OneParamValidator implements Validator {
 
-		protected final MethodParameter lastParameter;
+		private ParameterRequirement parameterRequirement;
 
-		public ParameterRequirement(ParameterValidator<?> anyOrderParameterValidator, MethodParameter lastParameter) {
-			this.validator = anyOrderParameterValidator;
-			this.lastParameter = lastParameter;
+		public OneParamValidator(ParameterRequirement param) {
+			this.parameterRequirement = param;
 		}
 
-		@SuppressWarnings("unchecked")
-		public final O optional() {
-			lastParameter.required = false;
-			return (O) validator;
+		public OneParamValidator optional() {
+			parameterRequirement.optional();
+			return this;
 		}
 
-		@SuppressWarnings("unchecked")
-		public final O required() {
-			return (O) validator;
+		public OneParamValidator multiple() {
+			parameterRequirement.multiple();
+			return this;
 		}
 
 		@Override
-		public final void validate(ExecutableElement executableElement, IsValid valid) {
-			validator.validate(executableElement, valid);
+		public void validate(ExecutableElement executableElement, IsValid valid) {
+			List<? extends VariableElement> parameters = executableElement.getParameters();
+			if (!parameterRequirement.multiple) {
+				if (parameterRequirement.required && parameters.size() != 1) {
+					invalidate(executableElement, valid);
+					return;
+				}
+				if (!parameterRequirement.required && parameters.size() > 1) {
+					invalidate(executableElement, valid);
+					return;
+				}
+			}
+
+			for (VariableElement parameter : parameters) {
+				if (!parameterRequirement.isSatisfied(parameter)) {
+					invalidate(executableElement, valid);
+					return;
+				}
+			}
 		}
 
+		protected void invalidate(ExecutableElement element, IsValid valid) {
+			annotationHelper.printAnnotationError(element, "%s can only have the following parameter: " + parameterRequirement);
+			valid.invalidate();
+		}
 	}
 
-	public static class ChainableParameterRequirement<O extends Validator> extends ParameterRequirement<O> implements ParameterValidator<ParameterRequirement<O>> {
-		public ChainableParameterRequirement(ParameterValidator<?> anyOrderParameterValidator, MethodParameter lastParameter) {
-			super(anyOrderParameterValidator, lastParameter);
-		}
+	private abstract class BaseParamValidator<V extends BaseParamValidator> implements Validator {
```

Comment:
I don't have any warning on this line with IntelliJ.
Is the warning on this line precisely ?